docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,237 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use AllReduce distribution strategy.
+
+## Motivation
+
+Now, users need to define the forward computation, loss function,
+optimizer and dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. Now ElasticDL only support
+Keras API to define the forward computation. It maybe not flexible for
+users to define complex models in CV or NLP. Sometimes, users need to
+customize the training loop to control the model iteration.
+
+If ElasticDL support customizing the training loop, users can use different
+deep learning library (e.g. TensorFlow, Pytorch) to define their training loop.
+ElasticDL only need to provide dynamic data partitioning for dataset and
+elastic AllReduce to merge gradients across workers.
+
+## The Training Loop of AllReduce
+
+The training loop of AllReduce contains two components:
+
+- Dataset to read data from disk and convert the data to tensor.
+- Calculate gradients and update the model.
+
+To support elastic training, the worker should read data from disk
+to create dataset according to the data task assigned by the master.
+So, ElasticDL can create a dataset for users and users only need to
+get batch data from the dataset to do forward and backward computation.
+
+```python
+def train(dataset):
+    for batch_data in dataset:
+        train_step(batch_data)
+```
+
+Using AllReduce distribution strategy, users need to merge gradients
+across workers before updating the local model during each step.
+
+```Python
+def train_step(batch_data):
+    loss = forward(batch_data)
+    gradients = backward(loss)
+    gradients = allreduce(gradients)
+    upate_model(gradients)
+```
+
+To support elastic training, the `allreduce` should be able to
+complete gradient combination even if the number of worker changes.
+
+## Elastic AllReduce to Merge Gradients in ElasticDL
+
+ElasticDL AllReduce is based on Horovod. Using Horovod, users only need to
+use Horovod APIs to wrap optimizer or tape to execute AllReduce in their
+training loop.
+
+```python
+def allreduce(model, optimizer, tape):
+    # Use Horovod DistributedGradientTape to wrap TensorFlow tape
+    tape = hvd.DistributedGradientTape(tape)
+    grads = tape.gradient(loss, model.trainable_variables)
+    optimizer.apply_gradients(
+        zip(grads, model.trainable_variables)
+    )
+```
+
+We can develop a decorator to wrap the gradient combination process
+using AllReduce. The decorator can support fault-tolerance and
+elastic training. If AllReduce fails, the decorator will query the
+master for a new AllReduce ring and retry to combine gradients
+across alive workers. And the decorator also queries the master
+for a new ring periodically in case that the master add new workers.
+
+```python
+@elastic_allreduce
```

Comment:
This code segment is confusing. If it is an example of how users to use elastic training with customized training loop,  the method definition should be consistent with the codes in the next section.

Reply:
I should delete the code segment.